server/src-lib/Hasura/GraphQL/Resolve/BoolExp.hs

```diff
@@ -158,13 +158,15 @@ parseColExp
 parseColExp nt n val = do
   fldInfo <- getFldInfo nt n
   case fldInfo of
-    Left pgColInfo -> do
+    RFPGColumn pgColInfo -> do
       opExps <- parseOpExps (pgiType pgColInfo) val
       return $ AVCol pgColInfo opExps
-    Right (RelationshipField relInfo _ _ permExp _)-> do
+    RFRelationship (RelationshipField relInfo _ _ permExp _)-> do
       relBoolExp <- parseBoolExp val
       return $ AVRel relInfo $ andAnnBoolExps relBoolExp $
         fmapAnnBoolExp partialSQLExpToUnresolvedVal permExp
+    RFComputedColumn _ -> throw500
+          "computed columns are not allowed in bool_exp"
```

Comment:
This is not impossible to implement. But it is not "straight forward" to implement. 
Will add this separately.

Reply:
I’m curious—what is the primary obstacle here? If a computed column is “just” a laterally-joined function, can’t we filter on it the same way we filter on other columns?

Reply:
```sql
-- Base root select
(
SELECT  * FROM "public"."author"
WHERE ... -- where clause goes here
)
```
Functions returning scalar values, we've to include the function call in `where` clause AND selection extractors
```sql
SELECT my_function("root") AS computed_field
FROM
(
  SELECT * FROM "public"."author" AS "table_row"
  WHERE my_function("public"."author") = 'some value'
) AS "root"
```
I'm planning to avoid two function calls do something like

```sql
SELECT my_function("root") AS computed_field
FROM
(
  SELECT * FROM "public"."author" 
) AS "root"
WHERE computed_field = 'some value'
```
This would potentially change some types in `Select/Internal.hs`. I'm planning to add this separately. Also, this PR is very huge. I thought it'd be prettier to add as a separate feature. :)

Reply:
> Functions returning scalar values, we've to include the function call in `where` clause AND selection extractors

Let’s say we have a query like this:

```graphql
query {
  posts(where: {is_public: {_eq: true}}) {
    id
    title
    computed_score
  }
}
```

I’d expect the SQL we generate to look sort of like this (aside from all the name-mangling we do, of course):

```sql
SELECT json_build_object(
    'id', posts.id,
    'title', posts.title,
    'computed_score', computed_score.value)
FROM posts
LEFT JOIN LATERAL
  (SELECT compute_post_score(posts.*))
  AS computed_score(value) ON true
WHERE posts.is_public = true
```

It seems to me that this really does just bring `computed_score` into exactly the same scope that `posts` is, so it should be very straightforward to handle this slightly modified query:

```diff
  query {
-   posts(where: {is_public: {_eq: true}}) {
+   posts(where: {is_public: {_eq: true}, computed_score: {_gt: 0}}) {
      id
      title
      computed_score
    }
  }
```

…since all we’d have to do is make this tiny change to the resulting SQL:

```diff
  SELECT json_build_object(
      'id', posts.id,
      'title', posts.title,
      'computed_score', computed_score.value)
  FROM posts
  LEFT JOIN LATERAL
    (SELECT compute_post_score(posts.*))
    AS computed_score(value) ON true
  WHERE posts.is_public = true
+   AND computed_score.value > 0
```

But your comment makes me think the SQL we generate isn’t like that after all, and I’m curious why not!

> Also, this PR is very huge. I thought it'd be prettier to add as a separate feature. :)

I appreciate the sentiment, but I think `throw500`s are always ugly. :)

Reply:
In my first comment

> This is not impossible to implement. But it is not "straight forward" to implement.

I didn't consider a case where things get complicated to implement computed fields in `where` expression. Let's say if the SQL function associated with a computed field has extra arguments other than table row type, then we've to pass them as input arguments. 

```sql
CREATE public.author_name_prefix(author_arg public.author, prefix text)
 RETURNS TEXT
 LANGUAGE sql
 STABLE
AS $function$
    SELECT prefix || ' ' || author_arg.name
$function$;
```
`name_with_prefix` is the computed field for `author_name_prefix` function
```graphql
query {
  author{
    name_with_prefix(args: {prefix: "Mr. "})
  }
}
```
The input for `name_with_prefix` in `where` will look like
```graphql
author(where: {name_with_prefix: {args: {prefix: "Mr. "}, _eq: ""}})
```
To achieve the above, we need to generate new set of boolean types
```graphql
input <table-name>_bool_exp {
  int_col: Int_comparison_exp
  string_col: String_comparison_exp
  computed_field: <table-name>_<computed-field-name>_comparison_exp
}
input <table-name>_<computed-field-name>_comparison_exp {
  args: <function_args>!
  _eq: <computed-field-type>
  # other comparison operators based on field type
}
```

Well, then it is possible to use different `args` values in `where` and in the actual selection
```graphql
query {
  author(where: {name_with_prefix: {args: {prefix: "Dr. "}, _eq: ""}}){
    name_with_prefix(args: {prefix: "Mr. "})
  }
}
```
Definitely we should make two function calls, one in `WHERE` clause and the other in extractors

```sql
SELECT author_name_prefix(author.*, 'Mr. ') as name_with_prefix FROM author WHERE author_name_prefix(author.*, 'Dr. ') = ' ';
```
Not over yet, we should also think about computed fields in permissions (`RQL` query). 

Boolean expression syntax in JSON may look like:-
```json
{
	"_where": {
		"computed_field": {
			"args": {"prefix": "Mr. "},
			"_eq": ""
		}
	}
}
```
`args` value can be static or values from session variables. 

I feel supporting computed fields in boolean expressions need fresh thinking and considerable code overhaul. This can go out as a separate feature/enhancement.

Reply:
> To achieve the above, we need to generate new set of boolean types

Ah, I see! Yes, I wasn’t thinking about the complication of needing to encode the function arguments in the boolean expressions differently, since input object fields don’t have “arguments” in the same way query fields do. Annoying! The issue of having different combinations of arguments in different places makes sense, too.

Sorry for missing the point for a while, but I appreciate the explanation.